lighthouse-cli/test/smokehouse/concurrent-mapper.js

```diff
@@ -0,0 +1,95 @@
+ */
+'use strict';
+
+/**
+ * A class that maintains a concurrency pool to coordinate many jobs that should
+ * only be run `concurrencyLimit` at a time.
+ * API inspired by http://bluebirdjs.com/docs/api/promise.map.html, but
+ * independent callers of `concurrentMap()` share the same concurrency limit.
+ */
+class ConcurrentMapper {
+  constructor() {
+    /** @type {Set<Promise<unknown>>} */
+    this._promisePool = new Set();
+
+    /**
+     * The limits of all currently running jobs. There will be duplicates.
+     * @type {Array<number>}
+     */
+    this._allConcurrencyLimits = [];
+  }
+
+  /**
+   * Returns whether there are fewer running jobs than the minimum current
+   * concurrency limit and the proposed new `concurrencyLimit`.
+   * @param {number} concurrencyLimit
+   */
+  _canRunMoreAtLimit(concurrencyLimit) {
+    return this._promisePool.size < concurrencyLimit &&
+        this._promisePool.size < Math.min(...this._allConcurrencyLimits);
+  }
+
+  /**
+   * Add a job to pool.
+   * @param {Promise<unknown>} job
+   * @param {number} concurrencyLimit
+   */
+  _addJob(job, concurrencyLimit) {
+    this._promisePool.add(job);
+    this._allConcurrencyLimits.push(concurrencyLimit);
+  }
+
+  /**
+   * Remove a job from pool.
+   * @param {Promise<unknown>} job
+   * @param {number} concurrencyLimit
+   */
+  _removeJob(job, concurrencyLimit) {
+    this._promisePool.delete(job);
+
+    const limitIndex = this._allConcurrencyLimits.indexOf(concurrencyLimit);
+    if (limitIndex === -1) {
+      throw new Error('No current limit found for finishing job');
+    }
+    this._allConcurrencyLimits.splice(limitIndex, 1);
+  }
+
+  /**
+   * Runs callbackfn on `values` in parallel, at a max of `concurrencyLimit` at
+   * a time across all callers on this instance. Resolves to an array of the
+   * results (for each caller separately) or rejects with the first rejected
+   * result.
+   * @template T, U
+   * @param {Array<T>} values
+   * @param {(value: T, index: number, array: Array<T>) => Promise<U>} callbackfn
+   * @param {number} concurrencyLimit
+   * @return {Promise<Array<U>>}
+   */
+  async map(values, callbackfn, concurrencyLimit) {
+    const result = [];
+
+    for (let i = 0; i < values.length; i++) {
+      // Wait until concurrencyLimit allows another run.
+      while (!this._canRunMoreAtLimit(concurrencyLimit)) {
+        // Unconditionally catch since we only care about our own failures
+        // (caught in the Promise.all below), not other callers.
+        await Promise.race(this._promisePool).catch(() => {});
+      }
+
+      // innerPromise removes itself from the pool and resolves on return from callback.
+      const innerPromise = callbackfn(values[i], i, values)
+        .finally(() => this._removeJob(innerPromise, concurrencyLimit));
+
+      this._addJob(innerPromise, concurrencyLimit);
```

Comment:
> (instead of waiting for all of A to finish before starting C), but still only running 5 at a time. The better utilization gives like a 40 second speed up on my machine (and could get us more if some of these serial tests can actually run in parallel).

Sorry should have been more clear. My assumption was that we would still be flattening the individual tests to have identical utilization, but I suppose there's just a tradeoff of complexity here in rearranging the smoke test objects vs. complex map calls. I'd lean toward streamlining the smoke test objects further over the rather complex multi-job map we've got here, but it's just smokehouse, so I'm fine.

Can we swap names though and have the straightforward one be `.map`? :) maybe like `.mapWithMultijobConcurrencyLimit` or `somethingMuchShorterButStillDifferentiated` :)

Reply:
> Can we swap names though and have the straightforward one be `.map`? :) maybe like `.mapWithMultijobConcurrencyLimit` or `somethingMuchShorterButStillDifferentiated` :)

this was done (`map()` is the static method, `pooledMap` is the shared one)

> My assumption was that we would still be flattening the individual tests to have identical utilization, but I suppose there's just a tradeoff of complexity here in rearranging the smoke test objects vs. complex map calls. I'd lean toward streamlining the smoke test objects further over the rather complex multi-job map we've got here, but it's just smokehouse, so I'm fine.

Overall, though, this is a point well taken. This class was fun to write but that doesn't count for much :)

The main downside of flattening would be a change to the smoke categories (`a11y`, `seo`, etc, which are super handy) unless switching to something like a tag-based system or, yeah, some heavy re-mapping.

If we could preserve the category behavior I'd be in favor of dropping all of this, but I think that's worth doing in a separate testDefn PR